tensorflow/lite/micro/kernels/cmsis_nn/softmax.cc

```diff
@@ -89,17 +135,111 @@ TfLiteStatus SoftmaxEval(TfLiteContext* context, TfLiteNode* node) {
+  const TfLiteEvalTensor* input = tflite::micro::GetEvalInput(context, node, 0);
+  TfLiteEvalTensor* output = tflite::micro::GetEvalOutput(context, node, 0);
+
+  TFLITE_DCHECK(node->user_data != nullptr);
+  const CMSISNNSoftmaxParams op_data =
+      *static_cast<const CMSISNNSoftmaxParams*>(node->user_data);
+
+  const cmsis_nn_softmax_lut_s16 softmax_params = {
+      .exp_lut = op_data.softmax_params.exp_lut,
+      .one_by_one_lut = op_data.softmax_params.one_over_one_plus_x_lut};
+
+  TFLITE_DCHECK_EQ(
+      arm_softmax_s16(tflite::micro::GetTensorData<int16_t>(input),
+                      op_data.num_rows, op_data.row_size,
+                      op_data.softmax_params.input_multiplier,
+                      op_data.softmax_params.input_left_shift, &softmax_params,
+                      tflite::micro::GetTensorData<int16_t>(output)),
+      ARM_MATH_SUCCESS);
+
+  return kTfLiteOk;
+}
+
 }  // namespace
 
 TfLiteRegistration Register_SOFTMAX() {
-  return {/*init=*/SoftmaxInit,
-          /*free=*/nullptr,
-          /*prepare=*/SoftmaxPrepare,
-          /*invoke=*/SoftmaxEval,
-          /*profiling_string=*/nullptr,
-          /*builtin_code=*/0,
-          /*custom_name=*/nullptr,
-          /*version=*/0};
+  softmax_registration.init = Init;
+  softmax_registration.free = nullptr;
+  softmax_registration.prepare = Prepare;
+  softmax_registration.invoke = SoftmaxEval;
+  softmax_registration.profiling_string = nullptr;
+  softmax_registration.builtin_code = 0;
+  softmax_registration.custom_name = nullptr;
+  softmax_registration.version = 0;
+  return softmax_registration;
+}
+
+TfLiteRegistration Register_SOFTMAX_INT8() {
+  softmax_registration.init = Init;
+  softmax_registration.free = nullptr;
+  softmax_registration.prepare = Prepare;
+  softmax_registration.invoke = SoftmaxEvalInt8;
+  softmax_registration.profiling_string = nullptr;
+  softmax_registration.builtin_code = 0;
+  softmax_registration.custom_name = nullptr;
+  softmax_registration.version = 0;
+  return softmax_registration;
+}
+
+TfLiteRegistration Register_SOFTMAX_INT8_INT16() {
+  softmax_registration.init = Init;
+  softmax_registration.free = nullptr;
+  softmax_registration.prepare = Prepare;
+  softmax_registration.invoke = SoftmaxEvalInt8_Int16;
+  softmax_registration.profiling_string = nullptr;
+  softmax_registration.builtin_code = 0;
+  softmax_registration.custom_name = nullptr;
+  softmax_registration.version = 0;
+  return softmax_registration;
+}
+
+TfLiteRegistration Register_SOFTMAX_INT16() {
+  softmax_registration.init = Init;
+  softmax_registration.free = nullptr;
+  softmax_registration.prepare = Prepare;
+  softmax_registration.invoke = SoftmaxEvalInt16;
+  softmax_registration.profiling_string = nullptr;
+  softmax_registration.builtin_code = 0;
+  softmax_registration.custom_name = nullptr;
+  softmax_registration.version = 0;
+  return softmax_registration;
```

Comment:
Is this change a style preference or is there a reason why we shouldn't use brace initialization here?

Reply:
The reason is that the symbol softmax_registration is needed in this context for the linker to drop unused symbols. There is a TODO(b/169801227). So it is needed to get the code size benefits of this quantization specific registration.

Reply:
@mansnils Thanks for sharing that historical context. We're moving towards a common API for kernel registration in #1084. The good thing is that the new approach also reduces code size (see PR description for numbers) compared to both brace initialization and static initialization. I suggest we make the same change here for `softmax` if #1084 is merged before this.

Reply:
Well yes, I tried with and without the new common kernel registration API for cmsis-nn/fully_conneted.cc. Then I tried with and without the int8 specific registration in person detect example: https://github.com/tensorflow/tflite-micro/blob/10b1223343303b45e6f62a433e25efb5f87958d7/tensorflow/lite/micro/examples/person_detection/main_functions.cc#L78
And the same code size difference is seen. I can't see the internals of TODO(b/169801227) but I guess it can be closed then.